File: patterns/python/sliding_window.py
class SlidingWindow:
    def find_max_average_brute_force(self, nums, k):
        max_avg = float('-inf')

        for i in range(len(nums) - k + 1):
            max_avg = max(max_avg, sum(nums[i:i + k]) / k)

        return max_avg

    def find_max_average_sliding_window(self, nums, k):
        sum_window = sum(nums[:k])
        max_sum = sum_window

        for i in range(k, len(nums)):
            sum_window += nums[i] - nums[i - k]
            max_sum = max(max_sum, sum_window)

        return max_sum / k

    def length_of_longest_substring_sliding_window(self, s):
        seen = set()
        max_length = left = 0

        for right in range(len(s)):
            while s[right] in seen:
                seen.remove(s[left])
                left += 1
            seen.add(s[right])
            max_length = max(max_length, right - left + 1)

        return max_length

    def length_of_longest_substring_sliding_window_frequency_array(self, s):
        freq = [0] * 128
        max_length = left = 0

        for right in range(len(s)):
            freq[ord(s[right])] += 1

            while freq[ord(s[right])] > 1:
                freq[ord(s[left])] -= 1
                left += 1

            max_length = max(max_length, right - left + 1)

        return max_length

    def maxSlidingWindow(self, nums, k):
        from collections import deque
        n = len(nums)
        max_deque = deque()
        z = [0] * (n - k + 1)

        for i in range(k):
            while max_deque and nums[max_deque[-1]] < nums[i]:
                max_deque.pop()
            max_deque.append(i)
        z[0] = nums[max_deque[0]]

        for i in range(k, n):
            while max_deque and nums[max_deque[-1]] < nums[i]:
                max_deque.pop()
            max_deque.append(i)
            if max_deque[0] == i - k:
                max_deque.popleft()
            z[i - k + 1] = nums[max_deque[0]]

        return z

File: patterns/c++/SlidingWindow.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <algorithm>

using namespace std;

class SlidingWindow {
public:
    double findMaxAverageBruteForce(vector<int>& nums, int k) {
        int n = nums.size();
        double maxAvg = INT_MIN;

        for (int i = 0; i <= n - k; i++) {
            int sum = 0;
            for (int j = i; j < i + k; j++) {
                sum += nums[j];
            }
            maxAvg = max(maxAvg, (double)sum / k);
        }
        return maxAvg;
    }

    double findMaxAverageSlidingWindow(vector<int>& nums, int k) {
        int n = nums.size();
        int sum = 0;

        for (int i = 0; i < k; i++) {
            sum += nums[i];
        }

        int maxSum = sum;

        for (int i = k; i < n; i++) {
            sum += nums[i];
            sum -= nums[i - k];
            maxSum = max(maxSum, sum);
        }

        return (double)maxSum / k;
    }

    int lengthOfLongestSubstringSlidingWindow(string s) {
        unordered_set<char> seen;
        int maxLength = 0, left = 0;

        for (int right = 0; right < s.size(); right++) {
            while (seen.count(s[right])) {
                seen.erase(s[left]);
                left++;
            }
            seen.insert(s[right]);
            maxLength = max(maxLength, right - left + 1);
        }
        return maxLength;
    }

    int lengthOfLongestSubstringSlidingWindowFrequencyArray(string s) {
        vector<int> freq(128, 0);
        int maxLength = 0, left = 0;

        for (int right = 0; right < s.size(); right++) {
            freq[s[right]]++;

            while (freq[s[right]] > 1) {
                freq[s[left]]--;
                left++;
            }

            maxLength = max(maxLength, right - left + 1);
        }
        return maxLength;
    }
};